### DISCUSSION TITLE
Non symmetric == operator in record subclasses overriding EqualityContract

### DISCUSSION THREAD

**Discussion by dariocc** [dotnet/csharplang#5586] [Q&A]
Hi!

After reading [c# 9 records proposal](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-9.0/records) I was experimenting with subclassing records and equality and I found that overriding `EqualityContract` breaks symmetry of the `==` operator:

```csharp
    public record Foo(string Value)
    {
    }

    public record Bar(string Value) : Foo(Value)
    {
        protected override Type EqualityContract => typeof(Foo);
    }
    
    [Test]
    public void TestFooBar()
    {
        Assert.That(new Foo("foo") == new Bar("foo"), Is.True); // Passes
        Assert.That(new Bar("foo") == new Foo("foo"), Is.True); // Fails!
    }
```

* Is this behavior intended?
* If so, doesn't it limit the usefulness of making it possible to override `EqualityContract`?

**Comment by agocke** [comment]
The purpose of the equality contract is to preserve symmetry -- if you override it, you may break symmetry. Sometimes, that may be what the user wants but you're right, it shouldn't be taken lightly!
